tuf/ngclient/fetcher.py

```diff
@@ -28,38 +28,66 @@ class FetcherInterface:
     __metaclass__ = abc.ABCMeta
 
     @abc.abstractmethod
-    def fetch(self, url: str) -> Iterator[bytes]:
+    def _fetch(self, url: str) -> Iterator[bytes]:
         """Fetches the contents of HTTP/HTTPS url from a remote server.
 
+        Implementations must raise FetcherHTTPError if they receive an
+        HTTP error code.
+
+        Implementations may raise any errors but the ones that are not
+        DownloadErrors will be wrapped in a DownloadError by fetch().
+
         Arguments:
             url: A URL string that represents a file location.
 
         Raises:
-            exceptions.SlowRetrievalError: A timeout occurs while receiving
-                data.
-            exceptions.FetcherHTTPError: An HTTP error code is received.
+            exceptions.FetcherHTTPError: An HTTP error code was received.
 
         Returns:
             A bytes iterator
         """
         raise NotImplementedError  # pragma: no cover
 
+    def fetch(self, url: str) -> Iterator[bytes]:
+        """Fetches the contents of HTTP/HTTPS url from a remote server.
+
+        Arguments:
+            url: A URL string that represents a file location.
+
+        Raises:
+            exceptions.DownloadError: An error occurred during download.
+            exceptions.FetcherHTTPError: An HTTP error code was received.
+
+        Returns:
+            A bytes iterator
+        """
+        # Ensure that fetch() only raises DownloadErrors, regardless of the
+        # fetcher implementation
+        try:
+            return self._fetch(url)
+        except Exception as e:
+            if isinstance(e, exceptions.DownloadError):
+                raise e
+            raise exceptions.DownloadError(f"Failed to download {url}") from e
+
     @contextmanager
     def download_file(self, url: str, max_length: int) -> Iterator[IO]:
         """Opens a connection to 'url' and downloads the content
         up to 'max_length'.
 
         Args:
-          url: a URL string that represents the location of the file.
-          max_length: an integer value representing the length of
-              the file or an upper bound.
+            url: a URL string that represents the location of the file.
+            max_length: an integer value representing the length of
+                the file or an upper bound.
 
         Raises:
-          exceptions.DownloadLengthMismatchError: downloaded bytes exceed
-            'max_length'.
+            exceptions.DownloadError: An error occurred during download.
+            exceptions.DownloadLengthMismatchError: downloaded bytes exceed
+                'max_length'.
+            exceptions.FetcherHTTPError: An HTTP error code was received.
```

Comment:
yeah, there's no great solutions here, I just picked the one I dislike least: I don't want to create new sub-errors just for the sake of it but the problem you mention does make the docstring less clear than it could be